In [4]:
import math
math.sqrt(2)

1.4142135623730951

In [10]:
X = 1

for i in range(5):
  X = (X  + 2/X) / 2
  print(X)

1.5
1.4166666666666665
1.4142156862745097
1.4142135623746899
1.414213562373095


In [11]:
print(round(3.1415926535, 2))
print(round(3.1415926535, 4))

3.14
3.1416


Pergunta 1
Sobre o codigo math.sqrt(2):Esse cálculo pode ser considerado um exemplo de cálculo direto?
Não. Embora para você, como programador, pareça uma execução "direta" (uma única linha de código), do ponto de vista do cálculo numérico, não é um método direto.

Justificativa:
Na análise numérica, um método direto é aquele que produz a solução exata do problema após um número finito e predeterminado de operações aritméticas (desconsiderando os erros de arredondamento da máquina). Como a raiz de 2 é um número irracional, não existe uma sequência finita de operações algébricas básicas que resulte no seu valor exato. Sob o capô do Python, a função embutida math.sqrt() utiliza algoritmos iterativos para convergir rapidamente e devolver uma aproximação extremamente precisa dentro do limite de casas decimais do computador.

Pergunta 2
Sobre o processo da raiz quadrada e o loop for:
O valor obtido melhora a cada iteração?

Sim. A cada iteração do loop, a nova aproximação (x) fica mais próxima do valor verdadeiro de raiz de 2 (aproximadamente 1.41421356...). O erro em relação à resposta real diminui progressivamente.
Esse processo pode ser classificado como cálculo direto ou iterativo? Explique.

É um cálculo iterativo.
Explicação: Métodos iterativos começam com um “chute” ou estimativa inicial (no seu código, x0 = 1) e aplicam repetidamente uma mesma fórmula ou regra (o bloco dentro do for). A saída de um passo torna-se a entrada do próximo. O objetivo não é encontrar a resposta em um número fixo de passos algébricos, mas sim gerar uma sequência de valores que se aproxima cada vez mais (converge) para a solução exata à medida que as iterações progridem.

Pergunta 3
Sobre os arredondamentos round(3.1415926535, 2) e round(3.1415926535, 4). Qual dos valores é “mais correto”?

Estritamente falando, o valor com 4 casas decimais (3.1416) é uma aproximação com maior exatidão do que o valor com 2 casas (3.14), pois o erro (a diferença entre o valor aproximado e o valor real de pi) é menor.
A resposta depende do problema considerado? Justifique.

Sim, totalmente. Em métodos numéricos e matemática aplicada, a quantidade de casas decimais que você escolhe reter depende da precisão exigida pelo problema real.

Justificativa: Todo arredondamento introduz um erro de truncamento. Se você estiver calculando a quantidade de material para fabricar um silo circular gigante ou projetando a trajetória de um satélite, usar apenas 3.14 introduziria um erro propagado massivo, tornando a resposta inadequada (neste caso, 3.1416 ou mais seria o correto). Por outro lado, se você estiver calculando a circunferência de um copo para um trabalho escolar simples, usar 3.14 é perfeitamente adequado, e usar 10 casas decimais seria um desperdício de processamento computacional sem ganho prático. O “correto” é o que satisfaz a tolerância de erro do projeto.